src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReadAheadWrappingStream.cs

```diff
@@ -0,0 +1,304 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// 
+
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Runtime;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ReadAheadWrappingStream : DelegatingStream
+    {
+        public const string ReadAheadWrappingStreamPropertyName = "ReadAheadWrappingStreamProperty";
+        private readonly int _readAheadBytes;
+        private readonly ArraySegment<byte>[] _buffers;
+        private Task<int> _previousCompletedReadTask;
+        private Task<int> _previousCompletedEnsureBufferedTask;
+        private bool _eof = false;
+
+        public ReadAheadWrappingStream(Stream innerStream, int readAheadBytes) : base(innerStream)
+        {
+            _readAheadBytes = readAheadBytes;
+            _buffers = new ArraySegment<byte>[2];
+            _buffers[0] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _buffers[1] = new ArraySegment<byte>(new byte[_readAheadBytes], 0, 0);
+            _previousCompletedReadTask = _previousCompletedEnsureBufferedTask = Task.FromResult(readAheadBytes);
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            if (_buffers[0].Count == 0 && _buffers[1].Count == 0)
+            {
+                if (_eof) // We've reached the end of the stream
+                {
+                    return Task.FromResult(0);
+                }
+
+                _previousCompletedReadTask = base.ReadAsync(buffer, offset, count, cancellationToken);
+                return _previousCompletedReadTask;
+            }
+
+            int bytesRead = ReadFromBuffer(buffer, offset, count);
+            if (bytesRead < count)
+            {
+                _previousCompletedReadTask = ReadAsyncAfterReadFromBuffer(bytesRead, buffer, offset, count, cancellationToken);
+                return _previousCompletedReadTask;
+            }
+
+            if (_previousCompletedReadTask.Result != bytesRead)
```

Comment:
Will _previousCompletedReadTask generally be completed by the time we ask for Result?  Should we use GetAwaiter().GetResult()?